test/e2e/storage/drivers/csi.go

```diff
@@ -0,0 +1,238 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+ * This file defines various csi volume test drivers for TestSuites.
```

Comment:
How can out-of-tree CSI drivers define their own instance of these shared tests?

They can't extend `drivers/csi.go`, because it is read-only for out-of-tree tests.

IMHO the in-tree CSI tests should be defined the same way. That way we know that the mechanism (whatever it is) works.

Reply:
This topic should be covered in https://github.com/kubernetes/kubernetes/issues/69819 .

As @msau42 suggested, we will be able to make this framework called from out-of-tree.
Simplest way would be:
  1. Out-of-tree driver developers define their driver in other places by the same way as test/e2e/storage/drivers/csi.go does, which means implement ```TestDriver``` interface defined in e2e/storage/drivers/base.go and its init function for their drivers.
  2. Also define similar code to test/e2e/storage/csi_volumes.go (after L98 is not necessary and ```csiTestDrivers``` need to be modified to include their driver's init code that is implemented in step 1 above).

We might improve step 2 not to duplicate the code, but its just about 100 steps of code including header and comments, so it should be enough simple.

Also, until moving csi driver test codes other than reference drivers are moved to out-of-tree, we still have a chance to add codes for out-of-tree drivers' e2e tests to in-tree. (As is intended in https://github.com/kubernetes/kubernetes/pull/67088 .)

Reply:
This sounds like it could work, thanks for explaining.